### PR TITLE
Add get_aws_accounts typed query

### PR DESCRIPTION
**Pull request by hemslo**
Add `get_aws_accounts` typed query using GraphQL query variables. This will be used in `terraform-tgw-attachments` integration.

Rely on https://github.com/app-sre/qontract-server/pull/200 for server side support.

### <samp>🤖 Generated by Copilot at 8c88cf3</samp>

This pull request adds typed queries and data classes for AWS accounts and their Terraform state. It uses qenerate to generate Python data classes from GraphQL definitions, and pydantic to validate and serialize the data. It also adds unit tests for the typed query functions. The main files affected are `reconcile/gql_definitions/terraform_tgw_attachments/aws_accounts.py` and `reconcile/typed_queries/terraform_tgw_attachments/aws_accounts.py`.

### <samp>🤖 Generated by Copilot at 8c88cf3</samp>

*  Add GraphQL fragment definitions and generated Python data classes for common fields and Terraform state of AWS accounts ([link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-041b9e2f9650f49eb656c41eae187b6d307733607219c1fdcf438979ef15c0faR1-R32), [link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-73edd9d10e4b60cd84b5557bf6daae0ebafbc9e28ac1cdf806f3481cb7fb1e95R1-R65), [link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-88acf0dd6e2ceca01a0671e1d65ca2bce2743d67be43d159bd5acb25fcf65a67R1-R11), [link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-156844dbf077ea17b4f80eb3db966f7c612c2fbfb204f872208298eac1833757R1-R38))
* Add GraphQL query definition and generated Python data classes for querying AWS accounts by name with the common fields and Terraform state fragments ([link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-6f027f305a09641fbeb3aa6566664e037132bb710be37e7764b844c4924bb937R1-R14), [link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-69928c976946acaae63d7bd7cdc93ad5af88574b59870a340e43107090772dd1R1-R118))
* Add convenience function `query` in `reconcile/gql_definitions/terraform_tgw_attachments/aws_accounts.py` that queries and parses the data into the data classes ([link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-69928c976946acaae63d7bd7cdc93ad5af88574b59870a340e43107090772dd1R1-R118))
* Add typed query function `get_aws_accounts` in `reconcile/typed_queries/terraform_tgw_attachments/aws_accounts.py` that takes a `GqlApi` instance and an optional name argument, and returns a list of `AWSAccountV1` instances using the `query` function ([link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-425ba1a6c7bd5a1e160fb790ef287b05bc63847a1491e934669f0dc8e71735b3R1-R18))
* Add unit tests for `get_aws_accounts` in `reconcile/test/test_typed_queries/terraform_tgw_attachments/test_get_aws_accounts.py` that use pytest fixtures and mock data classes and `GqlApi` instances, and assert the expected results and variables ([link](https://github.com/app-sre/qontract-reconcile/pull/3470/files?diff=unified&w=0#diff-044d783310977e23281246eab02d6397122cbbc6e9b673460e8eb736d6de971aR1-R101))

### CODE REVIEW

**Review by chassing — COMMENTED** [review]
TBH, I don't like such general-purpose queries. With our new GQL classes, it's much easier to fine-grain define all required attributes in a query instead of reusing existing ones. It would be better to define some important and commonly needed attributes as a fragment and reuse this fragment in gql queries defined for special use cases or integration needs.

**Comment by hemslo** [comment]
This is equivalent to [queries.get_aws_accounts](https://github.com/app-sre/qontract-reconcile/blob/344a8026c3cca7ad00f36f3846cfff90b7d61ba8/reconcile/queries.py#L543), but with typed query, there are many combinations of different usages, I'm afraid it may be too many similar pieces if we split them.

**Comment by chassing** [comment]
That is what I mean, do not follow the same pattern/path as with `queries.get_aws_account`. Having similar gql queries and similar GQL classes, but dedicated per integration, is completely fine, e.g. https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/gql_definitions/terraform_cloudflare_dns/app_interface_cloudflare_dns_settings.gql#L4 and https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/gql_definitions/terraform_cloudflare_users/app_interface_setting_cloudflare_and_vault.gql#L4

See the `gql_definitions` directory; each "newer" integration defines its own queries but reuses some fragment if it's worth.

**Comment by hemslo** [comment]
Cool, then I'll update this to only keep parts needed for `terraform-tgw-attachments`, it's hard to give it a proper name as the same structure is required by both `terraform-tgw-attachments` and `terraform-vpc-peerings`

**Comment by chassing** [comment]
Having 2 directories in `gql_definitions` (`terraform-tgw-attachments` and `terraform-vpc-peerings`) would be great. As I said, maybe it's worth having a proper fragment (in `fragments`) as well.

**Comment by hemslo** [comment]
@chassing I updated gql definitions and queries to fit `terraform-tgw-attachments`, is this the expected structure?

File: reconcile/test/test_typed_queries/terraform_tgw_attachments/test_get_aws_accounts.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+) -> None:
+    gql_api = gql_api_builder(aws_accounts_with_no_data.dict(by_alias=True))
+
+    result = get_aws_accounts(gql_api)
+
+    assert result == []
+
+
+def test_get_aws_accounts_with_name(
+    gql_api_builder: Callable[..., GqlApi],
```
[the remark this answers is not in the document]
Comment (hemslo) [reply reconcile/test/test_typed_queries/terraform_tgw_attachments/test_get_aws_accounts.py:86]: @chassing this is the result of this thread https://github.com/app-sre/qontract-reconcile/pull/3467#discussion_r1173829008.

TLDR:
* existing optional `query_func` doesn't have full test coverage when not using `mocker.patch`, the branch tested is not the same as the real call.
* inject query_func is very hard to make mypy happy, and tests become more complex
* change to `GqlApi` as it's easy to test, and not too hard to use, maybe easier if we have a dependency injection framework

**Review by chassing — APPROVED** [review]
LGTM! Thanks for all your effort! ❤️

**Pull request merged**